src/runtime/agas/addressing_service.cpp

```diff
@@ -3003,16 +3013,21 @@ addressing_service::send_refcnt_requests_async(
             stubs::primary_namespace::get_service_instance(raw)
           , naming::id_type::unmanaged);
 
-        requests[target].push_back(req);
+        (*requests)[target].push_back(req);
     }
 
     // send requests to all locality
-    requests_type::const_iterator end = requests.end();
-    for (requests_type::const_iterator it = requests.begin(); it != end; ++it)
+    requests_type::const_iterator end = requests->end();
+    for (requests_type::const_iterator it = requests->begin(); it != end; ++it)
     {
-        lazy_results.push_back(
+        // make sure requests vector goes out of scope only after the
+        // parcel-send operation is complete
+        future<std::vector<response> > f =
             stubs::primary_namespace::bulk_service_async(
-                (*it).first, (*it).second, action_priority_));
+                (*it).first, (*it).second,
```

Comment:
Even though the requests are stored in a shared pointer which is kept alive by the callback function, the requests are copied here. The actual requests that are part of the asynchronous communication are not being held alive, by the bound callback function.